### ISSUE TITLE
Black-and-White Gamma Correction: Difference between preview and output

### ISSUE THREAD

**Issue by Lawrence37** [Beep6581/RawTherapee#7159]
**Short description**
When L\*a\*b\* Adjustments is enabled with Avoid color shift set to Lab or Munsell only, the Black-and-White Gamma Correction makes the image colorful in the preview, but the output image is grayscale.

**Steps to reproduce**
1. Apply Neutral profile.
2. Enable L\*a\*b\* Adjustments. Change Avoid color shift to Lab or Munsell only.
3. Enable Black-and-White. Change any of the Gamma Correction values. Observe the color toning in the preview.
5. Save the image and open it in an image viewer. Observe the lack of color toning.

**Expected behavior**
The preview and output should match. I'm not sure if both are supposed to have color toning or if they should both stay black and white. However, staying black and white feels more natural and RawPedia does not mention any color toning effects of Gamma Correction.

**Comment by Benitoite** [comment]
```
9a245c1acb9bbbe5dd82ee3a4da4dcddc8903f35 is the first bad commit
commit 9a245c1acb9bbbe5dd82ee3a4da4dcddc8903f35
Author: Desmis <jdesmis@gmail.com>
Date:   Tue Jan 31 12:32:31 2023 +0100

    Avoid Color shift - Gamut and Munsell Review in RT - branch Munsellgamut (#6673)

    * Fixed numerous problems with gamut and Munsell in Local adjustments

    * change gamut-munsell in lab adjustmnts - gamut in ciecam

    * Improve XYZ colorimetry and tooltip

    * Change event - format code - change labels tooltips

    * Removed avoid_ and avoidmun_

    * Removed avoidcolorshift in labcurve

    * Push change proposed by Lawrence37 - compatibility with old pp3

 rtdata/languages/default  |  11 +-
 rtengine/improcfun.cc     | 482 +++++++++++++++++++++++++++-------------------
 rtengine/improcfun.h      |   2 +-
 rtengine/iplocallab.cc    | 148 +++++++++++---
 rtengine/procevents.h     |   6 +-
 rtengine/procparams.cc    |  51 +++--
 rtengine/procparams.h     |   6 +-
 rtengine/refreshmap.cc    |   3 +-
 rtgui/blackwhite.cc       |   3 +-
 rtgui/controlspotpanel.cc | 145 +++++++-------
 rtgui/controlspotpanel.h  |  19 +-
 rtgui/labcurve.cc         | 463 ++++++++++++++++++++++++--------------------
 rtgui/labcurve.h          |   9 +-
 rtgui/locallab.cc         |  56 +++++-
 rtgui/paramsedited.cc     |  30 ++-
 rtgui/paramsedited.h      |   5 +-
 16 files changed, 853 insertions(+), 586 deletions(-)
```

**Comment by Desmis** [comment]
@Lawrence37 @Benitoite 
I think it's a side effect of the "avoid color shift" changes.
I simply switched to gamutmap (XYZ) in the case of use (non none) of Black and white.

Branch blackwhite_avoid

**Comment by Benitoite** [comment]
@Desmis @Lawrence37 
Test of `blackwhite_avoid` is good, thanks for the fix.  Gamma correction correctly applied to input on preview and output is still good.

**Comment by Desmis** [comment]
@Lawrence37 @Benitoite 

I just create a Pull Request 
https://github.com/Beep6581/RawTherapee/pull/7165

**Comment by Lawrence37** [comment]
> I simply switched to gamutmap (XYZ) in the case of use (non none) of Black and white.

This sounds like a workaround. Does this give different results compared to a "proper" fix? My guess is that it does because of the different color spaces between avoid color shift and the black and white tool.

**Comment by Desmis** [comment]
@Lawrence37 

I don't know if we can talk about workaround, but the gamut control solutions are limited. In particular the "Lab" one which checks from RGB data, does not ensure the integrity of the data, that's why it was removed from "default".
 That's why I prefer "gamutmap".
So in conclusion, we must be able to improve "Gamut lab", but why do it. In the case of black and white, the use of gamutmap seems appropriate to me.

More generally, Lab poses a series of problems - sometimes poorly identified - which have led some to avoid it
https://rawpedia.rawtherapee.com/Toolchain_Pipeline#Colorimetry

But overall the use of Lab in RT is mastered and has the particularity of being simple to use

**Comment by Lawrence37** [comment]
Changing the avoid color shift method can be problematic because there is no indication to the user that it changed. Also, interactions between tools has a tendency to confuse users and make the code harder to maintain. I think there is a simple way to fix this issue without changing the avoid color shift method because it works correctly when saving the image.

**Comment by Desmis** [comment]
There are probably other solutions to solve this problem. I'll let you implement it.

As:
                if(params->blackwhite.enabled  && gamutmuns != 0) {//issue 7159
             //       gamutmuns = 2;//enable gamutmap XYZ - side effect due to change in gamut control
                    lnew->a[i][j] = 0.f;
                    lnew->b[i][j] = 0.f;
                }
or others....

Jacques

**Comment by Lawrence37** [comment]
I have success using the following patch:
```diff
diff --git a/rtengine/improcfun.cc b/rtengine/improcfun.cc
index 2e02a352c..b249809ae 100644
--- a/rtengine/improcfun.cc
+++ b/rtengine/improcfun.cc
@@ -4778,13 +4778,21 @@ void ImProcFunctions::chromiLuminanceCurve(PipetteBuffer *pipetteBuffer, int pW,
                     histLCurve[Lprov1 * histLFactor]++;
                 }
 
-                Chprov1 = sqrt(SQR(atmp) + SQR(btmp)) / 327.68f;
-
                 // labCurve.bwtoning option allows to decouple modulation of a & b curves by saturation
                 // with bwtoning enabled the net effect of a & b curves is visible
                 if (bwToning) {
                     atmp -= lold->a[i][j];
                     btmp -= lold->b[i][j];
+                    Chprov1 = sqrt(SQR(atmp) + SQR(btmp)) / 327.68f;
+                    if (Chprov1 == 0.f) {
+                        sincosval.x = 0.f;
+                        sincosval.y = 1.f;
+                    } else {
+                        sincosval.x = btmp / (327.68f * Chprov1);
+                        sincosval.y = atmp / (327.68f * Chprov1);
+                    }
+                } else {
+                    Chprov1 = sqrt(SQR(atmp) + SQR(btmp)) / 327.68f;
                 }
 
                 lnew->L[i][j] = Lprov1 * 327.68f;
```

The problem, I think, is that `Chprov1` and `sincosval` are not updated after `atmp` and `btmp` are. With Lab and Munsell only modes, `Chprov1` and `sincosval` are used. The other modes use `atmp` and `btmp`. That's why the issue only occurs with Lab and Munsell modes.

**Comment by Desmis** [comment]
@Lawrence37 

Well seen...It is a good change.

I just update the PR.

Thank you

Jacques

**Comment by Lawrence37** [comment]
Fix merged to branch `release-5.11`.
